src/System.Private.CoreLib/src/System/Threading/ManagedThreadId.cs

```diff
@@ -9,15 +9,171 @@
 // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
 
 using System;
-using System.Diagnostics.Contracts;
+using System.Diagnostics;
 using System.Runtime;
-using System.Runtime.InteropServices;
-using System.Security;
 
 namespace System.Threading
 {
     internal class ManagedThreadId
     {
+        //
+        // Binary tree used to keep track of active thread ids. Each node of the tree keeps track of 32 consecutive ids.
+        // Implemented as immutable collection to avoid locks. Each modification creates a new top level node.
+        // 
+        class ImmutableIdDispenser
+        {
+            private readonly ImmutableIdDispenser _left; // Child nodes
+            private readonly ImmutableIdDispenser _right;
+
+            private readonly int _used; // Number of ids tracked by this node and all its childs
+            private readonly int _size; // Maximum number of ids that can be tracked by this node and all its childs
+
+            private readonly uint _bitmap; // Bitmap of ids tracked by this node
+
+            private const int BitsPerNode = 32;
```

Comment:
I was not able to come up with anything simpler that has the desired perf characteristics. The problems are:

- It is important that the IdDispenser memory footprint shrinks to be proportional to the number of active threads. Array or linked list do not naturally shrink. A burst of many short-lived threads combined with a temporary hiccup on finalizer thread or longer GC should not cause the memory footprint to grow and *stay big* indefinitely. Here is my test for this situation: https://gist.github.com/jkotas/6fb00a78442c72440963 . In the current implementation, there will 10000+ elements in the linked list of recycled threadids stuck in it forever at the end of the test, even though the program had no more than just a couple dozen threads at any point.

- The required operations cannot be done on array or linked list lock-free, so there would need to be locks. A simple spin lock does not work well here as you have found, so it would need to be actual locks that scale well. We cannot use the regular locks here because of dependency cycle, so it would need to be some custom locks, perhaps PInvoke to pthread locks in CoreLibNative. But then you can just build the whole thing in CoreLibNative.

I think the next best solution - that has the desired perf characteristics - is platform specific thread id dispenser in C++. Maybe that would be better? Or if you would like to take a crack at building a simpler managed version, that is welcomed too - it is an interesting little algorithmic problem.